File: python/tests/test_transport_encryption_algorithm.py
"""v0.11 Task 2.4 (transport-spec §4.23): exercise the
``ENCRYPTION_ALGORITHM`` (0x1B) packet on
:class:`TransportWriter` + :class:`TransportReader`.

Wire layout per §4.23::

    algorithm_length:  uint16
    algorithm_utf8:    bytes[algorithm_length]

All multi-byte integers are LITTLE-ENDIAN (spec §1.7).

Python parity for Java's ``TransportEncryptionAlgorithmTest`` (commit
``530a5833``).

SPDX-License-Identifier: Apache-2.0
"""
from __future__ import annotations

import io
import struct
from pathlib import Path

from ttio.spectral_dataset import SpectralDataset
from ttio.transport.codec import TransportReader, TransportWriter
from ttio.transport.packets import PacketType, TRANSPORT_V0_11_FEATURE


def test_write_encryption_algorithm_emits_single_0x1B_packet() -> None:
    """Writer's low-level helper emits a single 0x1B packet whose
    payload matches §4.23 exactly."""
    out = io.BytesIO()
    with TransportWriter(out) as w:
        w.write_stream_header(
            format_version="1.2",
            title="enc-test",
            isa_investigation="isa",
            features=[],
            n_datasets=0,
        )
        w.write_encryption_algorithm("aes-256-gcm")
        w.write_end_of_stream()

    r = TransportReader(io.BytesIO(out.getvalue()))
    records = r.records_for_test()
    assert len(records) == 3, (
        f"expected StreamHeader + EncryptionAlgorithm + EndOfStream, "
        f"got {len(records)}"
    )
    assert records[1].header.packet_type == int(
        PacketType.ENCRYPTION_ALGORITHM
    )

    payload = records[1].payload
    (length,) = struct.unpack_from("<H", payload, 0)
    assert length == len(b"aes-256-gcm")
    algo = payload[2:2 + length].decode("utf-8")
    assert algo == "aes-256-gcm"
    assert 2 + length == len(payload), (
        "payload must contain only length + algorithm bytes"
    )


def test_write_dataset_emits_encryption_algorithm_when_encrypted(
    tmp_path: Path,
) -> None:
    """``write_dataset`` on an encrypted ``.tio`` emits exactly one
    ENCRYPTION_ALGORITHM packet (in the v0.11 prelude, before any
    reference groups per §5.4 ordering) AND sets the v0.11 feature
    flag in the StreamHeader."""
    src = tmp_path / "enc.tio"
    _build_encrypted_algorithm_only(src, "aes-256-gcm")
    tis = tmp_path / "enc.tis"

    with SpectralDataset.open(src) as ds:
        assert ds.is_encrypted, (
            "fixture precondition: dataset must be encrypted"
        )
        assert ds.encrypted_algorithm == "aes-256-gcm"
        with tis.open("wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    r = TransportReader(io.BytesIO(tis.read_bytes()))
    records = r.records_for_test()
    # StreamHeader features must include transport_v0_11.
    sh_payload = records[0].payload
    # Decode the StreamHeader to get its features list.
    from ttio.transport.codec import _decode_stream_header
    sh = _decode_stream_header(sh_payload)
    assert TRANSPORT_V0_11_FEATURE in sh["features"], (
        "StreamHeader must carry transport_v0_11 feature flag"
    )

    enc_count = 0
    for rec in records:
        if rec.header.packet_type == int(PacketType.ENCRYPTION_ALGORITHM):
            enc_count += 1
            (length,) = struct.unpack_from("<H", rec.payload, 0)
            algo = rec.payload[2:2 + length].decode("utf-8")
            assert algo == "aes-256-gcm"
    assert enc_count == 1, (
        f"write_dataset on encrypted .tio must emit exactly one "
        f"ENCRYPTION_ALGORITHM packet, got {enc_count}"
    )


def test_write_dataset_no_packet_when_not_encrypted(tmp_path: Path) -> None:
    """``write_dataset`` on a NON-encrypted ``.tio`` emits NO 0x1B
    packet."""
    src = tmp_path / "plain.tio"
    SpectralDataset.write_minimal(
        src,
        title="plain",
        isa_investigation_id="",
        runs={},
    )
    tis = tmp_path / "plain.tis"
    with SpectralDataset.open(src) as ds:
        assert not ds.is_encrypted, (
            "fixture precondition: dataset must not be encrypted"
        )
        with tis.open("wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    r = TransportReader(io.BytesIO(tis.read_bytes()))
    for rec in r.records_for_test():
        assert rec.header.packet_type != int(PacketType.ENCRYPTION_ALGORITHM), (
            "non-encrypted dataset must not emit ENCRYPTION_ALGORITHM"
        )


def test_encryption_algorithm_round_trips_via_write_dataset_read_to_dataset(
    tmp_path: Path,
) -> None:
    """End-to-end: writer emits ENCRYPTION_ALGORITHM, reader
    materialises it, the resulting on-disk ``.tio`` reports the same
    algorithm via ``is_encrypted`` + ``encrypted_algorithm``."""
    src = tmp_path / "src.tio"
    _build_encrypted_algorithm_only(src, "aes-256-gcm")
    tis = tmp_path / "src.tis"
    rt = tmp_path / "rt.tio"

    with SpectralDataset.open(src) as ds:
        with tis.open("wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    with TransportReader(io.BytesIO(tis.read_bytes())) as r:
        materialised = r.read_to_dataset(output_path=rt)
        materialised.close()

    with SpectralDataset.open(rt) as b:
        assert b.is_encrypted, (
            "round-tripped dataset must report is_encrypted == True"
        )
        assert b.encrypted_algorithm == "aes-256-gcm", (
            "round-tripped algorithm string must match source"
        )


def _build_encrypted_algorithm_only(target: Path, algorithm: str) -> None:
    """Build a minimal ``.tio`` whose root carries ``@encrypted`` =
    ``algorithm`` via the provider-level attribute setter. Mirrors
    Java's ``buildEncryptedAlgorithmOnly`` in
    ``TransportEncryptionAlgorithmTest``."""
    SpectralDataset.write_minimal(
        target,
        title="encryption_only",
        isa_investigation_id="",
        runs={},
    )
    # Re-open writable and stamp the @encrypted root attribute.
    with SpectralDataset.open(target, writable=True) as ds:
        ds.provider.root_group().set_attribute("encrypted", algorithm)

File: objc/Tests/TestTransportEncryptionAlgorithm.m
/*
 * TestTransportEncryptionAlgorithm.m — Task 3.4 of transport-spec v0.11.
 *
 * Exercises ENCRYPTION_ALGORITHM (0x1B) writer + reader. Wire layout
 * per transport-spec §4.23: `uint16 algorithm_length + bytes
 * algorithm_utf8[length]`, all LE.
 *
 * Cross-language parity:
 *   Java TransportEncryptionAlgorithmTest (commit 530a5833)
 *   Python tests/test_transport_encryption_algorithm.py (commit bf38bdc9)
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"

#import "Transport/TTIOTransportPacket.h"
#import "Transport/TTIOTransportReader.h"
#import "Transport/TTIOTransportReader+Internal.h"
#import "Transport/TTIOTransportWriter.h"
#import "Dataset/TTIOSpectralDataset.h"
#include <unistd.h>

static NSString *makeTempPathE(NSString *suffix)
{
    NSString *base = [NSString stringWithFormat:@"ttio_tr_enc_%d_%@.tio",
                       (int)getpid(), suffix];
    return [NSTemporaryDirectory()
        stringByAppendingPathComponent:base];
}

// ---------- helper: read u16 LE ----------

static uint16_t leU16(const uint8_t *b)
{
    return (uint16_t)((uint32_t)b[0] | ((uint32_t)b[1] << 8));
}

// ---------- writer low-level: single 0x1B packet ----------

static void testWriteEncryptionAlgorithmEmitsSingle0x1BPacket(void)
{
    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    NSError *err = nil;
    PASS([w writeStreamHeaderWithFormatVersion:@"1.2"
                                          title:@"enc-test"
                               isaInvestigation:@"isa"
                                       features:@[]
                                      nDatasets:0
                                          error:&err],
         "3.4: writeStreamHeader for low-level enc test");
    PASS([w writeEncryptionAlgorithm:@"AES-256-GCM" error:&err],
         "3.4: writeEncryptionAlgorithm emitted bytes");
    PASS([w writeEndOfStreamWithError:&err],
         "3.4: EndOfStream emitted");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records.count == 3,
         "3.4: expected StreamHeader + EncryptionAlgorithm + EndOfStream");
    PASS(records[1].header.packetType
            == TTIOTransportPacketEncryptionAlgorithm,
         "3.4: middle packet is ENCRYPTION_ALGORITHM (0x1B)");

    NSData *payload = records[1].payload;
    const uint8_t *bytes = payload.bytes;
    NSUInteger len = payload.length;
    PASS(len >= 2, "3.4: payload at least 2 bytes (uint16 length)");
    uint16_t algoLen = leU16(bytes);
    NSString *algo = [[NSString alloc] initWithBytes:&bytes[2]
                                                length:algoLen
                                              encoding:NSUTF8StringEncoding];
    PASS([algo isEqualToString:@"AES-256-GCM"],
         "3.4: algorithm decodes to AES-256-GCM");
    PASS((NSUInteger)(2 + algoLen) == len,
         "3.4: payload contains only length + algorithm bytes");
}

// ---------- writer: writeDataset on encrypted .tio emits 0x1B ----------

// Build a .tio whose root carries @encrypted = "aes-256-gcm" by writing
// a plain minimal dataset, then mutating the root attribute directly via
// TTIOHDF5File. Self-contained so we don't depend on TTIOEncryption.
#import "HDF5/TTIOHDF5File.h"
#import "HDF5/TTIOHDF5Group.h"

static NSString *buildEncryptedAlgorithmOnlyTio(NSString *base, NSError **err)
{
    NSString *path = makeTempPathE(base);
    unlink([path fileSystemRepresentation]);
    BOOL ok = [TTIOSpectralDataset writeMinimalToPath:path
                                                  title:@"enc-fixture"
                                     isaInvestigationId:@"ISA_ENC"
                                                 msRuns:@{}
                                        identifications:nil
                                        quantifications:nil
                                      provenanceRecords:nil
                                                  error:err];
    if (!ok) return nil;
    TTIOHDF5File *f = [TTIOHDF5File openAtPath:path error:err];
    if (!f) return nil;
    TTIOHDF5Group *root = [f rootGroup];
    if (![root setStringAttribute:@"encrypted"
                              value:@"aes-256-gcm"
                              error:err]) {
        [f close];
        return nil;
    }
    if (![f close]) return nil;
    return path;
}

static void testWriteDatasetEmitsEncryptionAlgorithmWhenEncrypted(void)
{
    NSError *err = nil;
    NSString *src = buildEncryptedAlgorithmOnlyTio(@"src_enc", &err);
    PASS(src != nil && err == nil, "3.4: built @encrypted fixture .tio");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil && ds.isEncrypted,
         "3.4: fixture dataset reports isEncrypted=YES");
    PASS([ds.encryptedAlgorithm isEqualToString:@"aes-256-gcm"],
         "3.4: fixture dataset reports encryptedAlgorithm=aes-256-gcm");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err],
         "3.4: writeDataset on encrypted .tio");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records != nil, "3.4: reader parsed encrypted-dataset stream");

    // StreamHeader features must carry transport_v0_11. The features
    // list is buried in the payload; quickest check is to scan as
    // UTF-8 bytes (Java + Python tests do the same).
    NSString *sh = [[NSString alloc] initWithData:records[0].payload
                                          encoding:NSUTF8StringEncoding];
    PASS(sh != nil && [sh containsString:@"transport_v0_11"],
         "3.4: StreamHeader carries transport_v0_11 feature flag");

    int encCount = 0;
    NSString *seenAlgo = nil;
    for (TTIOTransportPacketRecord *rec in records) {
        if (rec.header.packetType
                == TTIOTransportPacketEncryptionAlgorithm) {
            encCount++;
            const uint8_t *bytes = rec.payload.bytes;
            uint16_t algoLen = leU16(bytes);
            seenAlgo = [[NSString alloc] initWithBytes:&bytes[2]
                                                  length:algoLen
                                                encoding:NSUTF8StringEncoding];
        }
    }
    PASS(encCount == 1,
         "3.4: writeDataset on encrypted .tio emits exactly one 0x1B");
    PASS([seenAlgo isEqualToString:@"aes-256-gcm"],
         "3.4: emitted algorithm string matches source");

    [ds closeFile];
    unlink([src fileSystemRepresentation]);
}

// ---------- writer: writeDataset on plain .tio emits NO 0x1B ----------

static void testWriteDatasetNoPacketWhenNotEncrypted(void)
{
    NSError *err = nil;
    NSString *src = makeTempPathE(@"plain");
    unlink([src fileSystemRepresentation]);
    BOOL ok = [TTIOSpectralDataset writeMinimalToPath:src
                                                  title:@"plain"
                                     isaInvestigationId:@""
                                                 msRuns:@{}
                                        identifications:nil
                                        quantifications:nil
                                      provenanceRecords:nil
                                                  error:&err];
    PASS(ok && err == nil, "3.4 plain: minimal .tio created");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil && !ds.isEncrypted,
         "3.4 plain: dataset reports isEncrypted=NO");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err],
         "3.4 plain: writeDataset on plain .tio");

    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records != nil, "3.4 plain: reader parsed plain stream");

    BOOL sawEnc = NO;
    for (TTIOTransportPacketRecord *rec in records) {
        if (rec.header.packetType
                == TTIOTransportPacketEncryptionAlgorithm) {
            sawEnc = YES;
            break;
        }
    }
    PASS(!sawEnc,
         "3.4 plain: non-encrypted dataset emits no 0x1B packet");

    [ds closeFile];
    unlink([src fileSystemRepresentation]);
}

// ---------- end-to-end round-trip ----------

static void testEncryptionAlgorithmRoundTripsViaMaterialize(void)
{
    NSError *err = nil;
    NSString *src = buildEncryptedAlgorithmOnlyTio(@"src_rt", &err);
    PASS(src != nil, "3.4 rt: built encrypted source");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil, "3.4 rt: opened encrypted source");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err],
         "3.4 rt: writeDataset emitted stream");

    NSString *rt = makeTempPathE(@"rt");
    unlink([rt fileSystemRepresentation]);
    TTIOTransportReader *r = [[TTIOTransportReader alloc] initWithData:buf];
    PASS([r writeTtioToPath:rt error:&err] && err == nil,
         "3.4 rt: writeTtioToPath: materialised round-trip .tio");

    TTIOSpectralDataset *back =
        [TTIOSpectralDataset readFromFilePath:rt error:&err];
    PASS(back != nil, "3.4 rt: re-opened materialised .tio");
    PASS(back.isEncrypted,
         "3.4 rt: round-tripped dataset reports isEncrypted=YES");
    PASS([back.encryptedAlgorithm isEqualToString:@"aes-256-gcm"],
         "3.4 rt: round-tripped algorithm matches source");

    [ds closeFile];
    [back closeFile];
    unlink([src fileSystemRepresentation]);
    unlink([rt fileSystemRepresentation]);
}

void testTransportEncryptionAlgorithm(void);
void testTransportEncryptionAlgorithm(void)
{
    testWriteEncryptionAlgorithmEmitsSingle0x1BPacket();
    testWriteDatasetEmitsEncryptionAlgorithmWhenEncrypted();
    testWriteDatasetNoPacketWhenNotEncrypted();
    testEncryptionAlgorithmRoundTripsViaMaterialize();
}
